### PR TITLE
DOC: Add some entries for C types and macros

### PR DESCRIPTION
**Pull request by takanori-pskq**
Related to #13114.

### CODE REVIEW

File: doc/source/reference/c-api/types-and-structures.rst [hunk 1]
```diff
@@ -961,9 +959,14 @@ PyUFunc_Type and PyUFuncObject
...
        For each distinct core dimension, a set of ``UFUNC_CORE_DIM*`` flags
 
-       - :c:data:`UFUNC_CORE_DIM_CAN_IGNORE` if the dim name ends in ``?``
-       - :c:data:`UFUNC_CORE_DIM_SIZE_INFERRED` if the dim size will be
-         determined from the operands and not from a :ref:`frozen <frozen>` signature
+       .. c:macro:: UFUNC_CORE_DIM_IGNORE
+
+           if the dim name ends in ``?``
+
+       .. c:macro:: UFUNC_CORE_DIM_INFERRED
```
Comment (mattip) [inline doc/source/reference/c-api/types-and-structures.rst:966]: In line 956 this is `UFUNC_CORE_DIM_SIZE_INFERRED`.

File: doc/source/reference/c-api/types-and-structures.rst [hunk 2]
```diff
@@ -1438,3 +1441,5 @@ for completeness and assistance in understanding the code.
    ``arrayobject.h`` header. This type is not exposed to Python and
    could be replaced with a C-structure. As a Python type it takes
    advantage of reference- counted memory management.
+
+   .. c:type:: PyArrayMapIterObject
```
Comment (mattip) [inline doc/source/reference/c-api/types-and-structures.rst:1445]: This [looks a bit strange](https://16130-908607-gh.circle-artifacts.com/0/doc/build/html/reference/c-api/types-and-structures.html?highlight=types%20structures#c.PyArrayMapIterObject) without any descriptive text, but maybe that can be redone in a subsequent PR. But I think it should be dedented to mach the other types?

**Review by mattip — COMMENTED** [review]
Looks good overall. Just a couple of nits.

File: doc/source/reference/c-api/types-and-structures.rst (hunk 1 above)
Comment (takanori-pskq) [reply doc/source/reference/c-api/types-and-structures.rst]: Fixed.

File: doc/source/reference/c-api/types-and-structures.rst (hunk 2 above)
Comment (takanori-pskq) [reply doc/source/reference/c-api/types-and-structures.rst:1445]: For now `PyArrayMapIterObject` is not referenced anywhere else, so I think it is also possible to simply remove the reference above.

Comment (mattip) [reply doc/source/reference/c-api/types-and-structures.rst:1445]: sounds like a good idea.

**Comment by mattip** [comment]
Thanks @takanori-pskq

**Pull request merged**